crates/sui-archival/src/blob.rs

```diff
@@ -0,0 +1,549 @@
+use byteorder::{BigEndian, ByteOrder, ReadBytesExt};
+use crc::{Crc, CRC_32_ISCSI};
+use integer_encoding::VarInt;
+use integer_encoding::*;
+use num_enum::IntoPrimitive;
+use num_enum::TryFromPrimitive;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::fs::{read_dir, File, OpenOptions};
+use std::io::{BufReader, BufWriter, Read, Seek, SeekFrom, Write};
+use std::path::PathBuf;
+use std::{fs, mem};
+use tracing::info;
+use tracing::log::error;
+
+/// The following describes the format of a blob file, which is created in the `transactions/`,
+/// `/effects` or `/objects` directory of an archive. The maximum size per file is 32MB. Blobs (
+/// transactions, effects or objects) in the files are referenced from the index by uint64 composed
+/// of in-file offset (lower 4 bytes) and file sequence number (upper 4 bytes). Blobs are written
+/// out to local disk and shipped to remote storage optionally. A checkpoint tailer will subscribe
+/// to and continuously archive transactions and effects as blobs. For state snapshot, objects will
+/// be archived at the end of every epoch by scanning the checkpointed objects db table. One
+/// checkpoint directory could have data for multiple checkpoints in case of transactions and effects.
+/// For objects, a snapshot would be taken at the end of epoch and stored under the end of epoch
+/// checkpoint.
+/// Sui Archive Directory Layout
+///  - archive/
+///     - transactions/
+///       - checkpoint_000001/
+///         - 000001.blob
+///         - 000002.blob
+///         - 000003.blob
+///         - ...
+///         - 020000.blob
+///         - index
+///       - checkpoint_000100/
+///         - 000001.blob
+///     - effects/
+///       - checkpoint_000001/
+///         - 000001.blob
+///         - 000002.blob
+///         - ...
+///         - index
+///       - checkpoint_000050/
+///         - 000001.blob
+///         - 000002.blob
+///         - ....
+///         - index
+///     - objects/
+///       - checkpoint_100000/
+///         - 000001.blob
+///         - 000002.blob
+///         - ...
+///         - index
+///       - checkpoint_200000/
+///         - 000001.blob
+/// Blob File Disk Format
+///┌──────────────────────────────┐
+///│  magic(0xCAFE500D) <4 byte>  │
+///├──────────────────────────────┤
+///│    version(1) <1 byte>       │
+///├──────────────────────────────┤
+///│    padding(0) <3 byte>       │
+///├──────────────────────────────┤
+///│ ┌──────────────────────────┐ │
+///│ │         Blob 1           │ │
+///│ ├──────────────────────────┤ │
+///│ │          ...             │ │
+///│ ├──────────────────────────┤ │
+///│ │         Blob N           │ │
+///│ └──────────────────────────┘ │
+///└──────────────────────────────┘
+/// Blob
+///┌───────────────┬───────────────────┬──────────────┬────────────────┐
+///│ len <uvarint> │ encoding <1 byte> │ data <bytes> │ CRC32 <4 byte> │
```

Comment:
Varint saves us some bytes on disk if most blobs are smaller, so I imagine it has its usefulness (it is not any more complicated implementation wise but saves $$). sha3 is 20 bytes long. It is significantly bigger and doesn't provide any more value. We are looking for data integrity which crc32 is enough for. I wanted to build the blob format without an underlying assumption of the data that is being stored i.e. whether bit flips can be discovered by just reading the blobs (I am not even sure if it is true or we can still have silent corruptions for our current payloads) but having a checksum to know that for sure seems important. And one detail which is missing still is the index file. It will have file hashes (sha-3 like you suggested) of all blob files in the directory. But the index file itself could get corrupted, in which case we will need to regenerate the index by parsing all of the blob files while ensuring the data integrity of each blob in it.

Reply:
the etag is the checksum of the file as computed by the storage service - most object stores provide this. Its usually md5 for historical reasons.